src/Hazelcast.Net/Networking/ConnectAddressResolver.cs

```diff
@@ -0,0 +1,186 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Hazelcast.Core;
+using Hazelcast.Models;
+using Microsoft.Extensions.Logging;
+
+namespace Hazelcast.Networking
+{
+    /// <summary>
+    /// Resolves connect addresses for members by determining whether to use internal or public addresses.
+    /// </summary>
+    internal class ConnectAddressResolver
+    {
+        // TODO: consider making these options
+        private const int NumberOfMembersToCheck = 3;
+        private static readonly TimeSpan InternalAddressTimeout = TimeSpan.FromSeconds(1);
+        private static readonly TimeSpan PublicAddressTimeout = TimeSpan.FromSeconds(3);
+
+        private readonly NetworkingOptions _options;
+        private readonly ILogger _logger;
+
+        public ConnectAddressResolver(NetworkingOptions options, ILoggerFactory loggerFactory)
+        {
+            _options = options;
+            _logger = loggerFactory.CreateLogger<ConnectAddressResolver>();
+        }
+
+        /// <summary>
+        /// Determines whether to use public addresses.
+        /// </summary>
+        /// <param name="members">A collection of members.</param>
+        /// <returns><c>true</c> if public addresses must be used; otherwise <c>false</c>, indicating that internal addresses can be used.</returns>
+        public async Task<bool> DetermineUsePublicAddresses(IReadOnlyCollection<MemberInfo> members)
+        {
+            // if the user has specified its intention, respect it, otherwise try to decide
+            // automatically whether to use private or public addresses by trying to reach
+            // a few members
+
+            if (_options.UsePublicAddresses.HasValue)
+            {
+                var usePublicAddresses = _options.UsePublicAddresses.Value;
```

Comment:
Can be simplified a bit, but it's a matter of preference:
```suggestion
            if (_options.UsePublicAddresses is {} usePublicAddresses)
            {
```

Reply:
nice, going to use it